src/fs/mod.rs

```diff
@@ -0,0 +1,34 @@
+//! A high-level file system API for UEFI applications. It supports you to conveniently perform the
+//! following important operations:
+//! - read file to bytes
+//! - write bytes to file
+//! - create files
+//! - iterate directories and walk the file system tree
+//!
+//! Unlike UNIX-based file systems, there is no virtual file system. Thus, users perform actions on
+//! dedicated volumes: For example, the boot volume, such as a CD-rom, USB-stick, or a storage
+//! device. There are no symlinks or hard-links. Just plain paths, directories, and regular files.
+//! The way to access a volume is to open a [FileSystem].
+//!
+//! All paths are absolute and follow the FAT-like file system conventions for paths. Thus, there
+//! is no current working directory and path components like "." and ".." are not supported. In
+//! other words, the current working directory is always "/", i.e., the root, of the opened volume.
+//! This may change in the future but is currently sufficient.
+//!
+//! There are no `File` and `Path` abstractions that get publicly exported. Paths are provided as
+//! `&str` and validated internally. There are no `File` objects that are exposed to users.
+//!
+//! The difference to using the [SimpleFileSystemProtocol] directly is that the abstractions in this
+//! module take care of automatic release of resources and support you by returning data owned on
+//! the heap (such as file info). There is no synchronization as it is untypically that users
+//! bootstrap Application Processors (AP) during the UEFI stage, i.e., before hand-off to a kernel.
```

Comment:
Even though it's unlikely that an UEFI app will start the other processors, during the boot stage, sync is still a problem because of interrupts. You could have code which modifies an internal data structure, gets interrupted by an event, and then that event handler tries to access the same data structure (which is currently in an invalid, in progress state!). 

Fortunately, event handlers aren't exactly the usual place to do FS operations...

Reply:
Hm. I don't know how interrupts in UEFI are handled and where they are used. If an interrupt handler wants to access the `FileSystem` abstraction, while the lock is  contended, then.. what to do? Force an unlock the File System? @GabrielMajeri

Reply:
I think, it is the users responsibility to find a locking policy. A user might wrap the `FileSystem` implementation in a SpinLock or not. 

The documentation says that there is no out-of-the-box synchronization.